File: src/main/java/com/procurement/notice/controller/MainController.java
package com.procurement.notice.controller;

import com.fasterxml.jackson.databind.JsonNode;
import com.procurement.notice.model.bpe.ResponseDto;
import com.procurement.notice.service.MainService;
import java.time.LocalDateTime;
import javax.validation.Valid;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

@Validated
@RestController
@RequestMapping("/release")
public class MainController {

    private final MainService mainService;

    public MainController(final MainService mainService) {
        this.mainService = mainService;
    }

    @PostMapping
    public ResponseEntity<ResponseDto> createRelease(@RequestParam final String cpId,
                                                     @RequestParam(required = false) final String ocId,
                                                     @RequestParam final String stage,
                                                     @RequestParam(required = false) final String previousStage,
                                                     @RequestParam final String operation,
                                                     @RequestParam(required = false) final String phase,
                                                     @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                                                     @RequestParam(required = false) final LocalDateTime releaseDate,
                                                     @Valid @RequestBody final JsonNode data) {
        return new ResponseEntity<>(
                mainService.createRelease(cpId,
                        ocId,
                        stage,
                        previousStage,
                        operation,
                        phase,
                        releaseDate,
                        data),
                HttpStatus.CREATED);
    }
}

File: src/main/kotlin/com/procurement/notice/controller/MainController.kt
package com.procurement.notice.controller

import com.fasterxml.jackson.databind.JsonNode
import com.procurement.notice.model.bpe.ResponseDto
import com.procurement.notice.service.MainService
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime
import javax.validation.Valid

@Validated
@RestController
@RequestMapping("/release")
class MainController(private val mainService: MainService) {

    @PostMapping
    fun createRelease(@RequestParam cpId: String,
                      @RequestParam(required = false) ocId: String?,
                      @RequestParam stage: String,
                      @RequestParam(required = false) previousStage: String?,
                      @RequestParam operation: String,
                      @RequestParam(required = false) phase: String?,
                      @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                      @RequestParam releaseDate: LocalDateTime,
                      @Valid @RequestBody data: JsonNode): ResponseEntity<ResponseDto<*>> {
        return ResponseEntity(
                mainService.createRelease(
                        cpId = cpId,
                        ocId = ocId,
                        newStage = stage,
                        previousStage = previousStage,
                        operation = operation,
                        phase = phase,
                        releaseDate = releaseDate,
                        data = data),
                HttpStatus.CREATED)
    }
}
